### PR TITLE
[Notebook] Run Parameters Action UI Component 

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR is part 1 of 2. 
This focuses UI Components on getting the default parameters of the notebook via the parameter cell that is already in the notebook. Once I parse those, I use them as default placeholders for the user to replace if they desire.
Once the user inputs new parameters I parse those new values and format them to be placed in the same format to be consumed for our URI Parameterization, documented [here](https://github.com/microsoft/azuredatastudio/blob/main/extensions/notebook/src/protocol/notebookUriHandler.ts#L45).

The second part of this is calling the showNotebookDocument extensibility API within Core, which I am currently working on to move that into NotebookService. Once I have that I will be able to uncomment the code for openParameterizedNotebook as it is already set up properly.

Typical Use Case (without opening parameterized notebook):
RunParametersActionPart1

If Notebook has no Parameter Cell:
We notify them and direct them to Parameterization Documentation [here](https://docs.microsoft.com/en-us/sql/azure-data-studio/notebooks/notebooks-parameterization?view=sql-server-ver15). (Raden approved)
image

### CODE REVIEW

**Comment by VasuBhog** [comment]
Will be adding tests.

File: src/sql/media/icons/run-with-parameters.svg
Comment (halerankin) [inline src/sql/media/icons/run-with-parameters.svg:1]: I see you have a copy of this SVG in another location.
Please use exactly one SVG.

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts
Comment (halerankin) [inline src/sql/workbench/contrib/notebook/browser/notebook.component.ts:451]: Did you decide to have this feature available to users who don't have preview features enabled? If so, please make sure this icon presents like the others in this context so they are uniform in their presentation.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (halerankin) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:43]: +1 for consolidation
Thanks for shifting these two out of the blocks below!

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebook.component.ts:451]: For now, we can keep it in preview features only. So I will go ahead and delete this instance.

**Review by halerankin — COMMENTED** [review]
Left you some feedback.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (barbaravaldez) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:306]: I don't think you need this first `trim()`

Comment (barbaravaldez) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:306]: Also, I'm not sure if we want to take into account string values that could have the equal sign? 
Maybe splitting the string is not the best option.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:306]: The first trim is needed as the parameters will come in as ['x = 1 \n', 'y = 2'], as a result the first parameter when we parse will make the string have a new line. I think ensuring that both are trimmed ensures that there is no additional space added when adding to the map. 

Also, I think the scenario where values have equal sign would be an edge case but I could focus on splitting on the first instance of the = sign instead.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:320]: Consider adding a return after this in case more logic is added later outside the if's

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:327]: If the user cancels out of the quickpick this is going to be undefined - do you want to set that to the value at that point? (vs just leaving it as the original value)

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:330]: `to append to the URI`

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:332]: This will also ignore empty string values - is that intended?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:334]: Typically a little more readable if you use \`${key}=${pick.get(key)\`

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:337]: It's kind of odd to just increment the index once like this. Maybe instead just don't add/clear out any values which would be set to undefined and then just use string.join(..., '&') and prefix with ?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:334]: I feel like these query parsing/serializing things could be put in helper functions in our own uri.ts module in case anyone else wants to do the same kind of stuff. 

Also did you look around to see if there's any libraries we already use that have this kind of functionality built in? For example I see some people suggesting to use [UrlSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams)

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:352]: xml doc comments

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
Overall mostly looks fine - but it would be interesting to look into that UrlSearchParams first. That might save us a lot of effort doing our own manual parsing

File: src/sql/workbench/contrib/notebook/browser/notebook.css
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebook.css:185]: Since the feature is not gonna be there without preview features. I can take this out as well.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:327]: A user can click, or skip the first default parameter for example but enter the second parameter. If we keep it undefined then it will allow us to understand that the user only wants the second default parameter value to be updated and injected. (This is what the if condition within the next for loop checks)

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (corivera) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:277]: `toggleTooltip `on its own line? That would fit with the other arguments

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:277]: Sure can do that.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:306]: Parameter cells should only have variables and their assignments. Comments should be placed in text cells ahead of the code cell in most cases.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:272]: Sounds good.

File: src/sql/workbench/contrib/notebook/browser/notebook.css
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebook.css:66]: This was for the cell toolbar more actions feature that I split up with this PR as well. Will remove it as it's not needed for this PR.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:332]: Yes, and therefore utilize the default value from the original notebook for the new parameterized notebook.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:306]: Let's make sure that'd documented or something then - I assume there's nothing preventing someone from making a notebook and adding comments to the parameter cell.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:332]: What if users want to specifically have an empty string? I'm sure there's a use case for that. Not sure how we'd communicate that to the user though when they're filling in the values though...

Anyways guess I'll leave it up to you and design to decide what the best approach is here. But empty strings are not necessarily always invalid values so definitely consider that.

**Comment by VasuBhog** [comment]
After discussion with UX, we discussed what the best way to accept empty strings would be. Instead of using a placeholder visual for the quickInput, we use the actual value of the default parameter value and therefore if a user wants to set the parameter to an empty string then they must delete the original default parameter value. 

RunParametersActionUpdate

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:339]: What about languages that don't support ' as their string delimiters?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:347]: Why are you unescaping this?

1.  I would expect `openParameterizedNotebook` to take in a URI instead of just a string
2. Something else should be doing the unescaping, whenever you have something in URI form I would expect that it's escaped properly (and only unescape it when you extra out the query in the place that's going to use it)

File: src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts:261]: You don't have to comment this out - just having test.skip is enough to make it not run the test.

File: src/vs/workbench/services/configurationResolver/test/electron-browser/configurationResolverService.test.ts
Comment (Charles-Gagnon) [inline src/vs/workbench/services/configurationResolver/test/electron-browser/configurationResolverService.test.ts:646]: SQL CARBON EDIT

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:326]: I'd consider having `ignoreFocusLost` be true - something like this I can possibly see users wanting to go and check something else in the notebook or something and cause it to lose focus and having it cancel at that point would be kind of annoying.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
You have some hygiene errors now

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:331]: So right now it'd be required that the users include the string delimiters for string values as well - otherwise it'll break. Did you discuss this with design at all? I'm not sure that's going to be clear for people and so might cause some annoyances.

You could probably just follow up with this if you get some complaints - but I'd run it by some other people to get opinions on what to do here. It's tricky because you don't know for sure whether the param is supposed to be a string and so just blindly adding delimiters doesn't seem right either (even if it was a string originally maybe they want to enter a different param type now)

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:347]: I looked at this further - why are you even taking in both the URI and the filepath anyways? They both should be coming from the URI I would imagine

File: src/sql/workbench/contrib/notebook/browser/notebook.css
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebook.css:88]: The icon looks relatively smaller than the other ones on the toolbar so I increased the size so that it's a similar size to other icons, discussed this with Hale as well.

**Review by halerankin — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:339]: Currently, we are trying to make it generic but focused on Python kernel for parameterization. Most use cases for parameterization are through Python kernel. The other format would be supporting both ' and " as string delimiters.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:331]: I did run it by design, Raden and I discussed this. Users would have to include string delimiters for string values as we do not plan to do the validation at this moment. 

That is why I changed the setting from placeHolder to value in quickInput so that the user will know that for said parameter value it expects strings, as shown by the first parameter in the new gif I posted. We plan on addressing the concern however later as this ties in directly to the parameters validation aspect. For now, this will work.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:347]: So `openParameterizedNotebook` follows the same logic to how notebookURIHandler is setup for URI Parameterization. 

We must take the filePath as a string with the correct unescaped parameters in order for when we build the new uri from this filePath it will story the query parameters correctly in the uri.query. 
This would then be consumed properly on the notebookModel side as that is what we are able to retrieve from the uri. 
I need to unescape the uriParams.toString() as we discussed that this would be not correctly formed to how we would expect it - `Input.ipynb?x=1&y=2'`. 

The uri that is being passed in is of the currentNotebook which will get us the modelContents. The filePath is of the current notebook with the new parameters added to it which would then be parsed to create a new URI.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:347]: The other way to solely pass the URI would be to utilize the current URI and then set the uriParams query to the query. 
`x= unescape(uriParams.toString())
uri.with({query: x})` before passing it to `openParameterizedNotebook`.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:347]: Yup, that's exactly what I'm suggesting.

AFAIK we're kind of in unprecedented territory here - I don't think any of the current editors support extra parameters like this. But doing it this way makes sense to me since we already pass in a URI and this would keep things consistent with how the file URIs are passed around

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:339]: As long as that's expected. I will be curious to see how you go about solving this though in a generic way since there isn't really any way to tell how a language delimits its strings. Probably just have to go with a hardcoded mapping in the kernel specs or something I would imagine.

File: src/sql/workbench/contrib/notebook/browser/notebook.css
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebook.css:88]: mmm, that's definitely a smell that something isn't right here to me. If the icons are the same why wouldn't they be the same size in the UI? (changing the size of the SVG not affecting it makes sense to me though - the size of the SVG doesn't matter at all since we scale these to specific pixel height/widths anyways)

 But if you don't have time to investigate this now then ok - I'll probably take a look once you get this merged in.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebook.css:88]: Without the padding:
image

With the padding:
image

Just for reference. Can address in a later PR after investigating further.

File: src/vs/workbench/services/configurationResolver/test/electron-browser/configurationResolverService.test.ts
Comment (VasuBhog) [reply src/vs/workbench/services/configurationResolver/test/electron-browser/configurationResolverService.test.ts:646]: There is bad layering issue when calling this, so I will have to export this out to another file to be able to call it in notebookAction.test.ts

File: src/vs/workbench/services/configurationResolver/common/quickInputServiceMock.ts
Comment (Charles-Gagnon) [inline src/vs/workbench/services/configurationResolver/common/quickInputServiceMock.ts:1]: Don't just create new files in vs folder. If anything you should instead create this under the sql folder.

But as suggested elsewhere I would say it would be better to just move the test into electron-browser instead, which should fix the layering issue.

**Review by Charles-Gagnon — APPROVED** [review]

**Pull request merged**
